### PR TITLE
Improve API docs.

### PR DESCRIPTION
**Pull request by frankban**
Add links for all the endpoints.
Improve code formatting, and general clean up.
The result can be seen at  https://github.com/frankban/charmstore/blob/improve-api-doc/docs/API.md

### CODE REVIEW

**Comment by frankban** [comment]
Note: I'll update the endpoint documentation links in the code in a follow up branch.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/731/
Test PASSed.

File: docs/API.md [hunk 1]
```diff
@@ -3,47 +3,59 @@
...
+The charm store stores and indexes charms and bundles. A charm or bundle is
+referred to by a charm store id which can take one of the following two forms:
 
 * ~owner/series/name(-revision)
 * series/name(-revision)
 
 *Owner* is the name of the user that owns the charm.
-*Series* is one of a small number of known possible series for charms (currently just the Ubuntu series names) or the special name bundle to signify that the charm id refers to a charm bundle.
+*Series* is one of a small number of known possible series for charms
+(currently just the Ubuntu series names) or the special name bundle to signify
```
Comment (mitechie) [inline docs/API.md:15]: the special bundle part isn't true. Should we update this?

**Comment by mhilton** [comment]
:+1:

File: docs/API.md [hunk 2]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+The `Info` field is set when a request returns a "multiple errors" error code;
+currently the only two endpoints that can are `/meta` and `id/meta/any`.
+Each element in `Info` corresponds to an element in the PUT request, and holds
+the error for that element. See those endpoints for examples.
 
 ### Bulk requests and missing metadata
-There are two forms of "bulk" API request that can return information about several items at once. The meta/any endpoint (and some others) have a set of "include" flags that specify a bunch of metadata to return. The /meta endpoint has a set of "id" flags that specify a bunch of ids to return data on.
 
+There are two forms of "bulk" API request that can return information about
+several items at once. The `/meta/any` endpoint (and some others) have a set of
```
Comment (mitechie) [inline docs/API.md]: (along with some others)

File: docs/API.md [hunk 3]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+"include" flags that specify a bunch of metadata to return. The `/meta`
```
Comment (mitechie) [inline docs/API.md]: let's nix 'a bunch'

File: docs/API.md [hunk 4]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+endpoint has a set of "id" flags that specify a bunch of ids to return data on.
```
Comment (mitechie) [inline docs/API.md]: 'a bunch' to 'a set'

File: docs/API.md [hunk 5]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+#### POST id/archive
 
-
-Any additional elements attached to the `/charm` path retrieve the file from the charm or bundle's zip file. The `x-content-sha384` header field in the response will hold the hash checksum of the archive.
+This uploads the given charm or bundle in zip format.
 
 `POST id/archive?hash=sha384hash`
 
-This uploads the given charm or bundle in zip format. The id specified must specify the series and must not contain a revision number. The hash flag must specify the SHA384 hash of the uploaded archive in hexadecimal format. If the same content has already been uploaded, the response will return immediately without reading the entire body.
+The id specified must specify the series and must not contain a revision
```
Comment (mitechie) [inline docs/API.md:169]: is this true that it must not include a revision number?

Comment (mitechie) [inline docs/API.md:169]: sorry, this is POST ignore me.

File: docs/API.md [hunk 6]
```diff
@@ -1139,224 +1184,235 @@ type StatsCount struct {
...
-                Revision int
-        }
+```go
+type ResourcesRevision struct {
+        Revision int
+}
 ```
 
-`GET  id/resources/name.stream[-revision]/arch/filename`
+#### GET id/resources/name.stream[-revision]/arch/filename
```
Comment (mitechie) [inline docs/API.md]: I missed this. Do we have this implemented?

**Comment by mitechie** [comment]
:+1: with some minor tweaks. Thanks for the update. Good docs ftw!

File: docs/API.md (hunk 1 above)
Comment (rogpeppe) [inline docs/API.md:15]: What makes you think it's not true?

File: docs/API.md [hunk 7]
```diff
@@ -3,47 +3,59 @@
 The current live api lives at https://api.jujucharms.com/charmstore/v4
 
 ## Intro
-The charm store stores and indexes charms and bundles. A charm or bundle is referred to by a charm store id which can take one of the following two forms:
+
+The charm store stores and indexes charms and bundles. A charm or bundle is
+referred to by a charm store id which can take one of the following two forms:
 
 * ~owner/series/name(-revision)
```
Comment (rogpeppe) [inline docs/API.md]: can we italicise "owner", "series", "name" and "revision here and in the line below, please?

File: docs/API.md [hunk 8]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+response will hold the hash checksum of the archive.
 
-`GET id/archive/…`
+#### POST id/archive
 
-
-Any additional elements attached to the `/charm` path retrieve the file from the charm or bundle's zip file. The `x-content-sha384` header field in the response will hold the hash checksum of the archive.
+This uploads the given charm or bundle in zip format.
 
 `POST id/archive?hash=sha384hash`
```
Comment (rogpeppe) [inline docs/API.md]: italicise sha384hash

File: docs/API.md [hunk 9]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+Example response body:
 
-```
+```json
 {
     "Id": "precise/wordpress-24"
 }
 ```
 
-`DELETE id/archive`
```
Comment (rogpeppe) [inline docs/API.md:160]: italicise id

File: docs/API.md [hunk 10]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+#### DELETE id/archive
 
+This deletes the given charm or bundle with the given id. If the ID is not
+fully specified, the charm series or revisions are not resolved and the charm
+is not deleted. In order to delete the charm, the ID must include series as
+well as revisions. In order to delete all versions of the charm, use
+`/expand-id` and iterate on all elements in the result.
 
 ### Visual diagram
-`GET id/diagram.svg`
```
Comment (rogpeppe) [inline docs/API.md:166]: italicise id (and for all similar endpoints below)

File: docs/API.md [hunk 11]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
 
 This returns the README.
 
 ### Stats
 
+#### GET stats/counter/...
+
+This endpoint can be used to retrieve stats related to entities.
+
 `GET stats/counter/key[:key]...?[by=unit]&start=date][&stop=date][&list=1]`
```
Comment (rogpeppe) [inline docs/API.md]: italicise key, unit, date

File: docs/API.md [hunk 12]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
 
-The stats path allows the retrieval of counts of operations in a general way. A statistic is composed of an ordered tuple of keys: `kind:series:name:user`. Operations on the store increment counts associated with a specific tuple, determined by the operation and the charm being operated on.
+The stats path allows the retrieval of counts of operations in a general way. A
+statistic is composed of an ordered tuple of keys: `kind:series:name:user`.
```
Comment (rogpeppe) [inline docs/API.md]: italicise kind, series, name, user

File: docs/API.md [hunk 13]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+    "id-series",
+    "id-user",
+    "manifest",
+    "revision-info",
+    "stats",
+    "tags"
 ]
 ```
 
-`GET meta/$endpoint?id=$id0[&id=$id1...][$otherflags]`
```
Comment (rogpeppe) [inline docs/API.md:290]: To fit in with others, remove the $ signs and italicise id0, id1, otherflags

File: docs/API.md [hunk 14]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+#### GET meta/$endpoint
 
-This endpoint allows a user to query any number of IDs for metadata. This call is equivalent to calling `id/meta` for each id separately. The result holds an element for each id id in the request with the resulting metadata exactly as returned by `GET id/meta/$endpoint[?$otherflags]`. The map keys are the ids exactly as specified in the request, although they are resolved to fill in series and revision as usual when fetching the metadata. Any ids that are not found, or with non-relevant metadata, will be omitted.
+This endpoint allows a user to query any number of IDs for metadata.
 
-`map[string] interface{}`
+`GET meta/$endpoint?id=$id0[&id=$id1...][$otherflags]`
```
Comment (rogpeppe) [inline docs/API.md]: ditto

File: docs/API.md [hunk 15]
```diff
@@ -51,292 +63,350 @@ Note: this format is compatible with the error results used by juju-core. Curren
...
+            },
+            "extra-info/featured": true
         }
-  }
+    }
 }
 ```
 
-`PUT meta/endpoint`
+#### PUT meta/$endpoint
```
Comment (rogpeppe) [inline docs/API.md]: lose $, italicise endpoint.

File: docs/API.md (hunk 1 above)
Comment (mitechie) [inline docs/API.md:15]: @rogpeppe sorry you're right. We've not yet updated to the flat namespace ids on the charmstore api.

File: docs/API.md [hunk 16]
```diff
@@ -344,175 +414,184 @@ Example:
...
+                    "Code": "not found"
                 }
+            }
         }
+    }
+}
 ```
 
-`GET id/meta`
+#### GET id/meta
```
Comment (rogpeppe) [inline docs/API.md]: italicise id

File: docs/API.md [hunk 17]
```diff
@@ -344,175 +414,184 @@ Example:
...
+    "stats",
+    "tags"
 ]
 ```
 
-`GET id/meta/any?[include=meta[&include=meta...]]`
+#### GET id/meta/any
 
-The `meta/any` path returns requested metadata information on the given id. If the id is non-specific, the latest revision and preferred series for the id will be assumed.
+The `meta/any` path returns requested metadata information on the given id. If
```
Comment (rogpeppe) [inline docs/API.md:503]: This text no longer makes much sense without the metadata include
flags (what requested metadata information?)

File: docs/API.md [hunk 18]
```diff
@@ -344,175 +414,184 @@ Example:
...
+the id is non-specific, the latest revision and preferred series for the id
+will be assumed.
 
-Other metadata can be requested by specifying one or more `include` flags. The value of each meta must be the name of one of the path elements defined under the `/meta` path (for example: `charm-config`, `charm-meta`, `manifest`) and causes the desired metadata to be included in the Meta field, keyed by meta. If there is no metadata for the given meta path, the element will be omitted (for example, if bundle-specific data is requested for a charm id).
+`GET id/meta/any?[include=meta[&include=meta...]]`
```
Comment (rogpeppe) [inline docs/API.md]: italicise id, meta

File: docs/API.md [hunk 19]
```diff
@@ -344,175 +414,184 @@ Example:
...
+        "archive-size": {
+           "Size": 7580
+        },
+        "extra-info/featured": true
+	}
 }
 ```
 
-`PUT id/meta/any`
+#### PUT id/meta/any
```
Comment (rogpeppe) [inline docs/API.md]: italicise id

File: docs/API.md [hunk 20]
```diff
@@ -812,311 +884,284 @@ type Option struct {
...
+any bundles that contain the charm with the given id.
 
-```
-        []Bundle
-        type Bundle struct {
-                Id string
-                Meta map[string]interface{}                `json:",omitempty"`
-        }
-```
+`GET id/meta/bundles-containing[?include=meta[&include=meta…]]`
```
Comment (rogpeppe) [inline docs/API.md]: italicise id, meta

File: docs/API.md [hunk 21]
```diff
@@ -1139,224 +1184,235 @@ type StatsCount struct {
...
-                        “cs:trusty/wordpress-42”,
-                        “cs:trusty/wordpress-41”,
-                        “cs:trusty/wordpress-39”,
-                ]
-        }
-```
-
 
-`GET id/meta/id`
+#### GET id/meta/id
```
Comment (rogpeppe) [inline docs/API.md]: italicise the first occurrence of "id"

File: docs/API.md (hunk 6 above)
Comment (rogpeppe) [inline docs/API.md]: No. Juju-core never finalised their resources spec, so this is still in nascent form.

File: docs/API.md [hunk 22]
```diff
@@ -1139,224 +1184,235 @@ type StatsCount struct {
...
 
 ### Search
 
-Implementation document.
+#### GET search
+
+The `search` path searches within the latest version of charms and bundles
+within the store.
 
 `GET search[?text=text][&autocomplete=1][&filter=value…][&limit=limit][&skip=skip][&include=meta[&include=meta...]][&sort=field]`
```
Comment (rogpeppe) [inline docs/API.md]: italicise text, value, limit, skip, meta, field

File: docs/API.md [hunk 23]
```diff
@@ -1557,57 +1623,67 @@ type Log struct {
...
-
 ### Changes
-Each charmstore has a global feed for all new published charms and bundles. However, each entity in charmstore could be its own feed, and users would be able to see the changes in specific charms and bundles (probably also resources). Also, once “bundle loving” comes into place, we would provide information about charms and bundles being shared between users within same groups, etc.
-
 
-Beside filtering of information and limiting the number of returned values, we would also like to provide time limited queries. The client is responsible to track the “last checked” time and to provide proper time limited queries.
+Each charm store has a global feed for all new published charms and bundles.
+However, each entity in charm store could be its own feed, and users would be
+able to see the changes in specific charms and bundles (probably also
+resources). Also, once “bundle loving” comes into place, we would provide
```
Comment (rogpeppe) [inline docs/API.md]: Do we want to talk about this here?

**Comment by rogpeppe** [comment]
Thanks a lot for working on this.
LGTM with a bunch of italicisation suggestions to help the reader distinguish
between literal and symbolic text.

File: docs/API.md (hunk 2 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 3 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 4 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 6 above)
Comment (frankban) [inline docs/API.md]: Yeah, I specify that it's not yet implemented above.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/732/
Test PASSed.

File: docs/API.md (hunk 7 above)
Comment (frankban) [inline docs/API.md]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/733/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/734/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/735/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/736/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/737/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/738/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/739/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/740/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/741/
Test PASSed.

File: docs/API.md (hunk 20 above)
Comment (frankban) [inline docs/API.md]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/742/
Test PASSed.

File: docs/API.md (hunk 8 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 9 above)
Comment (frankban) [inline docs/API.md:160]: Done.

File: docs/API.md (hunk 10 above)
Comment (frankban) [inline docs/API.md:166]: Done.

File: docs/API.md (hunk 11 above)
Comment (frankban) [inline docs/API.md]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/743/
Test PASSed.

File: docs/API.md (hunk 12 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 13 above)
Comment (frankban) [inline docs/API.md:290]: Done.

File: docs/API.md (hunk 14 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 15 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 16 above)
Comment (frankban) [inline docs/API.md]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/744/
Test PASSed.

File: docs/API.md (hunk 17 above)
Comment (frankban) [inline docs/API.md:503]: Fixed.

File: docs/API.md (hunk 18 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 19 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 21 above)
Comment (frankban) [inline docs/API.md]: Done

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/745/
Test PASSed.

File: docs/API.md (hunk 22 above)
Comment (frankban) [inline docs/API.md]: Done.

File: docs/API.md (hunk 23 above)
Comment (frankban) [inline docs/API.md]: No, removed, good catch.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/746/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/747/
Test PASSed.

**Comment by frankban** [comment]
:shipit: 

Thanks for the reviews!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/748/
Test PASSed.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/749/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/750/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/751/
Test PASSed.

**Pull request merged**